joomla/joomla-cms
# [4.0] Proper cache invalidation of the static assets [.js/.css]

Pull Request for Issue # .

## The problem
For some years now Joomla has a `mediaversion` PHP function which essentially creates a hash from the Joomla version. This was/is essential the weird part at the end of all the js/css files (`?8567748356`) and the reason for its existence was/is to invalidate the browser cache (the filename is the same but somehow we need to tell the browser to re fetch the content because there's a new Joomla version so the file might have been changed). But the process is extremely inefficient because **FOR EVERY JOOMLA UPDATE WE INVALIDATE ALL THE STATIC ASSETS** whether there was a change or not!

## The solution
Since the introduction of Web Assets from @Fedik we have a registry where we can declaratively set some attributes. Amongst those attributes is also the `version` and this has been already mentioned by me a year ago when I was building the first truly SPA/PWA template: https://github.com/dgrammatiko/sloth/commit/f3151117a7a0f6fe2c90baa78fade1e25b0e449e Nobody picked the code to do a PR so here I am doing it myself 

### Summary of Changes
- Add a script that will add a version attribute in every asset in the joomla.assets.json (only in the media folder)
- Move some code to utils (DRY)
- Fix the bootstrap ES5 (the commonjs plugin was missing from the rollup config)


### Testing Instructions
- This PR needs CLI commands - NPM
- Apply this PR by git or by downloading the package
- Run `npm install` or `npm ci`
- Copy the file `media/system/joomla.asset.json` to `media/system/default-joomla.asset.json`
- Run the command `npm run versioning`
- Compare the 2 files and confirm that a `version ` attribute is added in every style/script asset
- Check that the actual hash in the file is used in Joomla


### Actual result BEFORE applying this Pull Request
```json
    {
      "name": "core",
      "type": "script",
      "class": "CoreAssetItem",
      "uri": "system/core.min.js"
    },
```

### Expected result AFTER applying this Pull Request
```json
    {
      "name": "core",
      "type": "script",
      "class": "CoreAssetItem",
      "uri": "system/core.min.js",
      "version": "8f1a05a7e6e6fad83d22a4a5b67bafbb04ddc875"
    },
```


### Documentation Changes Required

This basically as the gzip command affects only the release leader or anyone that will create a Joomla release but I guess this needs to be added in the steps for creating a release

@wilsonge 

### TODO
- Create the appropriate `Joomla.assets.json` files for all the modules
- Create the appropriate `Joomla.assets.json` files for all the plugins
